### PR TITLE
Expose that Guice module configure() method is called multiple times

### PR DESCRIPTION
**Pull request by geminicaprograms**
### TestNG Version
`7.3.0` (also reproducible `7.0.0` and version built out of `49070a7c`)

### Expected behaviour
When suite is executed modules (parent and test) should have *configure()* method called only once.
That behaviour is observed in `6.14.3`.

### Actual behaviour
When suite is called *configure()* method is called multiple times for both parent and test modules.

### Is the issue reproducible on runner?

- [ ] Shell
- [*] Maven
- [ ] Gradle
- [ ] Ant
- [*] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

I haven't tested other runners but I believe it will be reproducible there too.

### CODE REVIEW

**Comment by geminicaprograms** [comment]
The test in question (`issue188.IssueTest.testSuiteLevelParallelMode`) is flaky in my local env - it succeeds always when called in isolation from Eclipse but fails randomly when called from the command line. Is this sth normal or I should keep looking?

**Comment by krmahadevan** [comment]
@geminicaprograms - yes the test that you called out is a flaky one. I have attempted at fixing it via my PR https://github.com/cbeust/testng/pull/2424 (The PR is pending merge since it has some review comments that I need to address but havent found time to get around it due to stuff at work). So you can ignore that failing test.

**Comment by krmahadevan** [comment]
@juherr @cbeust - I am not well versed yet with Guice DI. Can you please help with reviewing this PR ?

**Comment by krmahadevan** [comment]
@geminicaprograms - Can you please help add an entry into `CHANGES.txt` referring this issue in it ?

File: src/main/java/org/testng/IInjectorFactory.java [hunk 1]
```diff
@@ -11,11 +12,12 @@
...
   /**
+   * @param parent - Parent {@link com.google.inject.Injector} instance that was built with parent injector
    * @param stage - A {@link Stage} object that defines the appropriate stage
    * @param modules - An array of {@link Module}
-   * @return - An {@link com.google.inject.Inject} instance that can be used to perform dependency
+   * @return - An {@link com.google.inject.Injector} instance that can be used to perform dependency
    * injection.
    */
-  Injector getInjector(Stage stage, Module... modules);
+  Injector getInjector(@Nullable Injector parent, Stage stage, Module... modules);
```
Comment (juherr) [inline src/main/java/org/testng/IInjectorFactory.java:21]: We can't change the public API easily.
Please deprecate the previous method first and add a default behavior for the new one.

**Review by juherr — CHANGES_REQUESTED** [review]
The change looks good to me if the tests are OK.
But we cannot change the public API and previous suites must still work until the next removal of deprecated (maybe for the next, unplanned, major version)

File: src/main/java/org/testng/IInjectorFactory.java (hunk 1 above)
Comment (geminicaprograms) [reply src/main/java/org/testng/IInjectorFactory.java]: Done. Note that it cannot be marked as @FunctionalInterface anymore as it contains more methods now...

**Comment by geminicaprograms** [comment]
> @geminicaprograms - Can you please help add an entry into `CHANGES.txt` referring this issue in it ?

Added as part of the latest commit ;)

File: src/test/java/test/guice/issue2427/Test1.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.testng.annotations.Guice;
+import org.testng.annotations.Test;
+import test.guice.issue2427.modules.TestModuleOne;
+
+@Guice(modules = {TestModuleOne.class})
+@Test()
+public class Test1 {
+  public void shouldInstatiateModulesOnlyOnce() {
+    System.out.println(String.format("Test %s called", getClass().getSimpleName()));
```
Comment (juherr) [inline src/test/java/test/guice/issue2427/Test1.java:11]: Please, remove the syso from the tests

File: src/main/java/org/testng/IInjectorFactory.java (hunk 1 above)
Comment (juherr) [reply src/main/java/org/testng/IInjectorFactory.java]: Right. If you prefer, a new interface could be used instead.

File: src/test/java/test/guice/issue2427/Test1.java (hunk 2 above)
Comment (geminicaprograms) [reply src/test/java/test/guice/issue2427/Test1.java]: Done.

File: src/main/java/org/testng/IInjectorFactory.java (hunk 1 above)
Comment (geminicaprograms) [reply src/main/java/org/testng/IInjectorFactory.java]: Yeah, was wondering about going into that direction as removing @FunctionalInterface is somewhat invasive although TBH it is not used that way in the existing code hence there is no big risk unless someone uses it that way in their code (can't think why would they ;))... anyway I have no strong opinion on any solution and will follow the lead here ;) So what is your take on that?

Comment (juherr) [reply src/main/java/org/testng/IInjectorFactory.java]: I don't have any strong opinion too.

@krmahadevan WDYT?

Comment (krmahadevan) [reply src/main/java/org/testng/IInjectorFactory.java]: @juherr - I believe that we will ONLY be temporarily revoking the `FunctionalInterface` stature of this interface till we get through with this deprecation strategy and remove off the original method. So it should be fine if we change the stature of this interface to a regular interface. I am fine with this change.

**Comment by geminicaprograms** [comment]
AFAICT it fails only on `issue188.IssueTest.testSuiteLevelParallelMode` test (which is AFAIK OK). Should we go ahead and merge it? ;)

**Review by krmahadevan — APPROVED** [review]

**Review by juherr — APPROVED** [review]

**Pull request merged**